requester/tracer.go

```diff
@@ -0,0 +1,85 @@
+package requester
+
+import (
+	"crypto/tls"
+	"net/http"
+	"net/http/httptrace"
+	"time"
+)
+
+// Event is a stage of an outgoing http request associated with a timestamp.
+type Event struct {
+	Name string        `json:"name"`
+	Time time.Duration `json:"time"`
+}
+
+// tracer is a http.RoundTripper to be used as a http.Transport
+// that records the events of an outgoing http request.
+type tracer struct {
+	start  time.Time
+	events []Event
+}
+
+// RoundTrip implements http.RoundTripper. It attaches the client trace
+// to the request context and calls http.DefaultTransport.RoundTrip
+// with the new created request.
+func (p *tracer) RoundTrip(r *http.Request) (*http.Response, error) {
+	ctx := httptrace.WithClientTrace(r.Context(), p.trace())
+	return http.DefaultTransport.RoundTrip(r.WithContext(ctx))
+}
+
+// trace returns a http.ClientTrace that timestamps and records the events
+// of an outgoint http requests.
+func (p *tracer) trace() *httptrace.ClientTrace {
+	return &httptrace.ClientTrace{
+		GetConn: func(string) {
+			p.start = time.Now()
+			p.addEvent("GetConn")
+		},
+		DNSStart: func(httptrace.DNSStartInfo) {
+			p.addEvent("DNSStart")
+		},
+		DNSDone: func(httptrace.DNSDoneInfo) {
+			p.addEvent("DNSDone")
+		},
+		ConnectStart: func(string, string) {
+			p.addEvent("ConnectStart")
+		},
+		ConnectDone: func(string, string, error) {
+			p.addEvent("ConnectDone")
+		},
+		GotConn: func(httptrace.GotConnInfo) {
+			p.addEvent("GotConn")
+		},
+		TLSHandshakeStart: func() {
+			p.addEvent("TLSHandshakeStart")
+		},
+		TLSHandshakeDone: func(tls.ConnectionState, error) {
+			p.addEvent("TLSHandshakeDone")
+		},
+
+		WroteHeaders: func() {
+			p.addEvent("WroteHeaders")
+		},
+		WroteRequest: func(httptrace.WroteRequestInfo) {
+			p.addEvent("WroteRequest")
+		},
+		GotFirstResponseByte: func() {
+			p.addEvent("GotFirstResponseByte")
+		},
+		PutIdleConn: func(error) {
+			p.addEvent("PutIdleConn")
+		},
```

Comment:
[question] It took me some time to understand why we have parameters we don't use.
It is just because it is in the signature of the hook methods, right?
And it is provided **just in case** we wanted to use it in the method called at this time?

Reply:
Yes exactly! The signatures are specific to the hooks and we have to comply with them (even if we don't actually use them).

Reply:
Question: is it better to completely avoid the param name or explicitly discard it with `_`:
- `func(string) {...}` vs
- `func(_ string) {...}`
I know that `_` is the way to discard so its obvious but I'd like your thoughts on that.

Reply:
`_` is necessary if one wants a ignore some parameter(s) amongst others:
```go
// unexpected
func Do(name string, int, isCool bool) {}
                  // ^^^ this is a bool param named "int"

// ok
func Do(name string, _ int, isCool bool) {}
```

Being unnecessary in our case, and parameters being there only to comply with the required signature, I prefer the more concise version to this:
```go
ConnectDone: func(_, _ string, _ error) {
    p.addEvent("ConnectDone")
},
```